src/intptrcast.rs

```diff
@@ -46,9 +46,23 @@ pub struct GlobalStateInner {
     provenance_mode: ProvenanceMode,
 }
 
-impl VisitTags for GlobalStateInner {
-    fn visit_tags(&self, _visit: &mut dyn FnMut(BorTag)) {
-        // Nothing to visit here.
+impl VisitProvenance for GlobalStateInner {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        let GlobalStateInner {
+            int_to_ptr_map: _,
+            base_addr: _,
+            exposed,
+            next_base_addr: _,
+            provenance_mode: _,
+        } = self;
+        // Though base_addr and int_to_ptr_map contain AllocIds, we do not want to visit them.
+        // We're trying to remove unused elements from base_addr, so visiting it would prevent
+        // removing anything. int_to_ptr_map is managed by free_alloc_id, and entries in it do not
+        // actually make allocation base addresses reachable so we don't need to visit it.
+        // But exposed tags do make base addresses reachable.
+        for id in exposed {
```

Comment:
There's no point in visiting exposed since it contains only live allocations, and those have already been added.
At least that should be true, we remove the allocation from `exposed` in `free_alloc_id`... but currently we don't check that it is live in `expose_ptr` *oops*

Similarly, the argument for not visiting `int_to_ptr_map` should be that it only contains live allocations.

Only base_addr needs a different argument: that map is only ever relevant when we *have* a pointer of given `AllocId` and need to know its base address; if the AllocId exists nowhere else we can remove it here.

Reply:
> but currently we don't check that it is live in `expose_ptr` _oops_

This sounds like a good thing to add a check for; but `MonoHashMap::contains_key` takes `&mut self` so the call is tricky to pull off here. I think the receiver should be changed to `&self`, but that would require changing the trait... I'll do that PR to the main repo.

Reply:
I am confused, can't we just add a liveness check trivially around [here](https://github.com/rust-lang/miri/blob/622a4d35b4a3f2404caa6834429d1ff452f5b27d/src/intptrcast.rs#L186)?

Reply:
Maybe I'm just doing it wrong, but I tried to do `ecx.machine.alloc_map().contains_key`.

Reply:
No, that's a *super* low-level function, don't use it.

`ecx.get_alloc_info(alloc_id)` returns `AllocKind::Dead` for dead allocations and something else for live allocations; you can use that.

Reply:
Of course, there are even other checks just like that in this file.